src/lib/backend_fapi.c

```diff
@@ -0,0 +1,490 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+
+#include "config.h"
+#include "backend_fapi.h"
+#include "emitter.h"
+#include "parser.h"
+#include <tss2/tss2_fapi.h>
+
+FAPI_CONTEXT *fctx = NULL;
+int objectids = 0;
+
+CK_RV backend_fapi_init(void) {
+    if (fctx != NULL) {
+        LOGW("Backend FAPI already initialized.");
+        return CKR_OK;
+    }
+    LOGV("Calling Fapi_Initialize");
+    TSS2_RC rc = Fapi_Initialize(&fctx, NULL);
+    if (rc) {
+        LOGW("Could not initialize FAPI");
+        return CKR_GENERAL_ERROR;
+    }
+    return CKR_OK;
+}
+
+CK_RV backend_fapi_destroy(void) {
+    LOGV("Calling Fapi_Finalize");
+    Fapi_Finalize(&fctx);
+    return CKR_OK;
+}
+
+CK_RV backend_fapi_ctx_new(token *t) {
+    t->fapi.ctx = fctx;
+    return CKR_OK;
+}
+
+void backend_fapi_ctx_free(token *t) {
+    (void)(t);
+}
+
+
+#define PREFIX "/HS/SRK/tpm2-pkcs11-token-"
+
+static char * tss_path_from_id(unsigned id, const char *type) {
+    char *path = malloc(strlen(PREFIX) + strlen(type) + 1 + 8 + 1);
+    if (!path)
+        return NULL;
+
+    sprintf(&path[0], PREFIX "%s-%08x", type, id);
+
+    return path;
+}
+
+/** Create a new token in fapi backend.
+ *
+ * See backend_create_token_seal()
+ */
+CK_RV backend_fapi_create_token_seal(token *t, const twist hexwrappingkey,
+                       const twist newauth, const twist newsalthex) {
+    TSS2_RC rc;
+    uint8_t *tpm2bPublic, *tpm2bPrivate, *appdata;
+    size_t tpm2bPublicSize, tpm2bPrivateSize, appdata_len;
+    twist pub, priv;
```

Comment:
Generally speaking, unless their is goto scope issues, this projects coding style is to declare variables nearest to use and also not to do the list of types. One variable declaration per line. This applies here and throughout.